### PR TITLE
[Refactor] New play mechanism

### PR DESCRIPTION
**Pull request by gnabyl**
This is a big refactor, which re-implements song queue and a new (better) mechanism to play songs consecutively.
Names changes:
- SongQueue => QueueManager
- QueueManager = Map<string, QueueObject>
- Each QueueObject holds information about connections and player.

This new mechanism is done using onStateChange event handler so it's supposed to be a more quality code.

### CODE REVIEW

**Comment by gnabyl** [comment]
Surprisingly good doc about audio here https://discordjs.github.io/voice/modules.html
Sorry for saying that discord's doc is shit

File: package.json [hunk 1]
```diff
@@ -3,10 +3,10 @@
   "version": "1.0.0",
   "description": "A discord music bot.",
   "main": "dist/index.js",
-  "engines" : {
```
Comment (amtcultist) [inline package.json:6]: weird spaces in package.json

File: src/commands/play.ts [hunk 2]
```diff
@@ -1,122 +1,61 @@
...
-  }
-  if (
-    queue.getQueue(guild.id).player._state.status === AudioPlayerStatus.Idle
-  ) {
-    play(guild, queue);
-  }
-}
+  queue.addTrack(track);
+  await interaction.editReply(`Track **${track.title}** added to the queue!`);
+}
```
Comment (amtcultist) [inline src/commands/play.ts]: eof

File: src/model/track-data.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+ */
+export interface TrackData {
+  title: string;
+  author: string;
+  url: string;
+  duration: string;
+  onStart: () => void;
+  onFinish: () => void;
+  onError: (error: Error) => void;
+}
```
Comment (amtcultist) [inline src/model/track-data.ts]: eof
all properties are require ?

File: src/utils/handlers.ts [hunk 4]
```diff
@@ -61,7 +62,7 @@ export class Handlers {
       case Commands.SKIP:
         this.lastCommand.set(interaction.guildId, interaction.user.id, {
           command: Commands.SKIP,
-          results: await executeSkip(interaction, this.queue),
```
Comment (amtcultist) [inline src/utils/handlers.ts:64]: good use of DI

File: src/utils/queue-object.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  private playNext() {
+    // If the queue is locked (already being processed), is empty, or the audio player is already playing something, return
+    if (this.isLocked || this.audioPlayer.state.status !== AudioPlayerStatus.Idle || this.songs.length === 0) {
+      return;
+    }
+    // Lock the queue
+    this.isLocked = true;
+
+    // Take the first item from the queue
+    const nextTrack = this.songs.shift();
```
Comment (amtcultist) [inline src/utils/queue-object.ts:127]: wait we remove the song from the queue to next ? 
Gonna be hard with the loop

**Review by amtcultist — APPROVED** [review]
only the shift, others are pretty decent

File: src/model/track-data.ts (hunk 3 above)
Comment (gnabyl) [reply src/model/track-data.ts]: Yes I guess?

File: src/utils/queue-object.ts (hunk 5 above)
Comment (gnabyl) [reply src/utils/queue-object.ts:127]: that's right, maybe we should have a pointer that moves, not shifting the queue :/

**Pull request merged**